experimental/parser/parse_expr.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/internal/taxa"
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/token"
+)
+
+// parseExpr attempts to parse a full expression.
+//
+// May return nil if parsing completely fails.
+func parseExpr(p *parser, c *token.Cursor, where taxa.Place) ast.ExprAny {
+	return parseExprInfix(p, c, where, ast.ExprAny{}, 0)
+}
+
+// parseExprInfix parses an infix expression.
+//
+// prec is the precedence; higher values mean tighter binding. This function calls itself
+// with higher (or equal) precedence values.
+func parseExprInfix(p *parser, c *token.Cursor, where taxa.Place, lhs ast.ExprAny, prec int) ast.ExprAny {
+	if lhs.Nil() {
+		lhs = parseExprPrefix(p, c, where)
+		if lhs.Nil() || c.Done() {
+			// Recursing will result where duplicate diagnostics.
+			return lhs
+		}
+	}
+
+	next := peekTokenExpr(p, c)
+	switch prec {
+	case 0:
+		if where.Subject == taxa.Array || where.Subject == taxa.Dict {
+			switch next.Text() {
+			case "=": // Allow equals signs, which are usually a mistake.
+				p.Errorf("unexpected `=` where expression").With(
+					report.Snippetf(next, "help: replace this with `:`"),
+					report.Note("top-level `option` assignment uses `=`, not `:`"),
```

Comment:
The way this is worded, it almost looks like it _wants_ an equals sign. Maybe instead something like ``"array and message literals use `:`, not `=`"``.